File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
import sys
from collections import deque

def convert_maps(maps):
    graph = []
    for map in maps:
        tmp_list = []
        for word in map:
            if word != 'X':
                tmp_list.append(int(word))
            else:
                tmp_list.append(-1)
        graph.append(tmp_list)
    return graph

def check_availability(graph):
    for i in graph:
        for j in i:
            if j != -1:
                return True
    return False
        

def solution(maps):
    answer = []
    graph = convert_maps(maps)
    if not check_availability(graph):
        return [-1]
    n = len(graph)
    m = len(graph[0])
    visited = [[False] * m for i in range(n)]
    
    dir_x = [1, 0, -1, 0]
    dir_y = [0, 1, 0, -1]
    
    for i in range(n):
        for j in range(m):
            if graph[i][j] != -1 and not visited[i][j]:
                queue = deque()
                queue.append([i, j])
                visited[i][j] = True
                days = graph[i][j]
                
                while queue:
                    cur = queue.popleft()
                    cur_x = cur[0]
                    cur_y = cur[1]
                    
                    for k in range(4):
                        new_x = cur_x + dir_x[k]
                        new_y = cur_y + dir_y[k]
                        if 0 <= new_x < n and 0 <= new_y < m and not visited[new_x][new_y] and graph[new_x][new_y] != -1:
                            days += graph[new_x][new_y]
                            queue.append([new_x, new_y])
                            visited[new_x][new_y] = True
                answer.append(days)
    
    answer.sort()
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.*;

class Solution {
    public boolean checkX(String[] maps) {
        int m = maps.length;
        int n = maps[0].length();
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char curChar = maps[i].charAt(j);
                if (curChar != 'X') {
                    return true;
                }
            }
        }
        return false;
    }
    
    public int[][] makeGraph(String[] maps) {
        int m = maps.length;
        int n = maps[0].length();
        int[][] graph = new int[m][n];
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char curChar = maps[i].charAt(j);
                if (curChar == 'X') {
                    graph[i][j] = 0;
                } else {
                    graph[i][j] = (int)curChar - '0';
                }
            }
        }
        return graph;
    }
    
    public boolean[][] makeVisited(int m, int n) {
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                visited[i][j] = false;
            }
        }
        return visited;
    }
    
    public int[] solution(String[] maps) {
        int[] answer = {};
        ArrayList<Integer> answerList = new ArrayList<>();
        
        // 지낼 수 있는 땅이 없을때
        if (!checkX(maps)) {
            answer = new int[1];
            answer[0] = -1;
            return answer;
        }
        
        int m = maps.length;
        int n = maps[0].length();
        
        int[][] graph = makeGraph(maps);
        boolean[][] visited = makeVisited(m, n);
        
        int[] dirX = {1, 0, -1, 0};
        int[] dirY = {0, 1, 0, -1};
        
        for (int sx = 0; sx < m; sx++) {
            for (int sy = 0; sy < n; sy++) {
                if (graph[sx][sy] != 0 && !visited[sx][sy]) {
                    Deque<Node> queue = new ArrayDeque<>();
                    Node start = new Node(sx, sy);
                    visited[sx][sy] = true;
                    queue.add(start);

                    int cnt = graph[sx][sy];
                    while (!queue.isEmpty()) {
                        Node curNode = queue.pollFirst();
                        int curX = curNode.x;
                        int curY = curNode.y;


                        for (int i = 0; i < 4; i++) {
                            int newX = curX + dirX[i];
                            int newY = curY + dirY[i];

                            if (newX >= 0 && newX < m && newY >= 0 && newY < n && !visited[newX][newY] && graph[newX][newY] != 0) {
                                cnt += graph[newX][newY];
                                visited[newX][newY] = true;
                                queue.add(new Node(newX, newY));
                            }
                        } 
                    }
                    answerList.add(cnt);
                }           
            }
        }
        Collections.sort(answerList);
        int size = answerList.size();
        answer = new int[size];
        for (int i = 0; i < size; i++) {
            answer[i] = answerList.get(i);
        }
        return answer;
    }
}

class Node {
    int x;
    int y;
    Node(int x, int y) {
        this.x = x;
        this.y = y;
    }
}